### PR TITLE
chore: add user flag for withdraw and burn

### PR DESCRIPTION
**Pull request by fcavazzoli**
# Description

## Changes: 
 - Add flags for burning and withdrawing.
 - When flagged user tries to double Burn or Withdraw, the returned error is `Err(TxError::Other("String")`

Now, one user can only perform one burn or one withdraw at a time. In case of dip20Proxy, the restriction is applied by token which means that one user is allowed to burn multiple tokens at the same time as long they are different canisters

### CODE REVIEW

File: eth_bridge/ic/src/eth_proxy/eth_proxy.did [hunk 1]
```diff
@@ -11,6 +11,7 @@ type Result_2 = variant { Ok; Err : text };
 type TxError = variant {
   InsufficientAllowance;
   InsufficientBalance;
+  MultipleTokenTx;
```
Comment (b0xtch) [inline eth_bridge/ic/src/eth_proxy/eth_proxy.did:14]: Let's keep parity in the `TxError` and use the `Other(String)`.

File: eth_bridge/ic/src/eth_proxy/src/api/burn.rs [hunk 2]
```diff
@@ -24,6 +24,11 @@ async fn burn(eth_addr: EthereumAddr, amount: Nat) -> TxReceipt {
         )));
     }
 
+    match STATE.with(|s| s.set_user_flag(caller, TxFlag::Burning)) {
```
Comment (b0xtch) [inline eth_bridge/ic/src/eth_proxy/src/api/burn.rs:27]: if let here might be cleaner

File: eth_bridge/ic/src/eth_proxy/src/api/withdraw.rs [hunk 3]
```diff
@@ -35,6 +35,11 @@ pub async fn withdraw(eth_addr: EthereumAddr, _amount: Nat) -> TxReceipt {
     let eth_addr_hex = WETH_ADDRESS_ETH.trim_start_matches("0x");
     let weth_eth_addr_pid = Principal::from_slice(&hex::decode(eth_addr_hex).unwrap());
 
+    match STATE.with(|s| s.set_user_flag(caller, TxFlag::Withdrawing)) {
```
Comment (b0xtch) [inline eth_bridge/ic/src/eth_proxy/src/api/withdraw.rs:38]: if let here might be cleaner

File: eth_bridge/ic/src/eth_proxy/src/common/types.rs [hunk 4]
```diff
@@ -103,6 +113,7 @@ pub enum TxError {
     Unauthorized,
     LedgerTrap,
     AmountTooSmall,
+    MultipleTokenTx,
```
Comment (b0xtch) [inline eth_bridge/ic/src/eth_proxy/src/common/types.rs:116]: Let's keep parity in the `TxError` and use the `Other(String)`.

File: magic_bridge/ic/src/dip20_proxy/src/api/burn.rs [hunk 5]
```diff
@@ -42,6 +42,12 @@ async fn burn(
     let erc20_addr_hex = ERC20_ADDRESS_ETH.trim_start_matches("0x");
     let erc20_addr_pid = Principal::from_slice(&hex::decode(erc20_addr_hex).unwrap());
 
+    // One user cannot make multiple tx at the same time for the same token
+    match STATE.with(|s| s.set_user_flag(caller, token_id, TxFlag::Burning)) {
```
Comment (b0xtch) [inline magic_bridge/ic/src/dip20_proxy/src/api/burn.rs:46]: if let here might be cleaner

**Review by b0xtch — APPROVED** [review]
Mint

File: eth_bridge/ic/src/eth_proxy/src/proxy.rs [hunk 6]
```diff
@@ -117,6 +117,31 @@ impl ProxyState {
...
+        Ok(())
+    }
+
+    pub fn get_user_flag(&self, user: Principal) -> Option<TxFlag> {
+        let flag = if let Some(flag) = self.user_actions.borrow().get(&user) {
+            flag.to_owned()
+        } else {
+            return None;
+        };
+        return Some(flag);
```
[the remark this answers is not in the document]
Comment (fcavazzoli) [reply eth_bridge/ic/src/eth_proxy/src/proxy.rs]: Done 👍

**Pull request merged**